### PR TITLE
Async startup of the SSHD server

### PR DESCRIPTION
**Pull request by olivergondza**
- Fix race condition in `SSHD.start()`.
- Do not let Jenkins to wait until `SSHD` is fully up.
  - I have seen launch stuck for minutes while running parallel `jenkins-test-harness` tests as it was waiting for system entropy to launch a service I do not care for on the first place.
- Some javadoc fixes while on it.

### CODE REVIEW

File: src/main/java/org/jenkinsci/main/modules/sshd/SSHD.java [hunk 1]
```diff
@@ -200,7 +207,15 @@ public static SSHD get() {
 
     @Initializer(after= InitMilestone.JOB_LOADED,fatal=false)
     public static void init() throws IOException, InterruptedException {
-        get().start();
+        MasterComputer.threadPoolForRemoting.submit(new Runnable() {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/main/modules/sshd/SSHD.java]: Log something like "Scheduling SSHD startup" on the FINE level?

File: src/main/java/org/jenkinsci/main/modules/sshd/SSHD.java [hunk 2]
```diff
@@ -200,7 +207,15 @@ public static SSHD get() {
...
+            @Override public void run() {
+                try {
+                    get().start();
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/main/modules/sshd/SSHD.java]: Some logging before and after may be also useful for diagnostic purposes

File: src/main/java/org/jenkinsci/main/modules/sshd/SSHD.java [hunk 3]
```diff
@@ -73,7 +74,7 @@ public int getPort() {
     /**
      * Gets the current TCP/IP port that this daemon is running with.
      *
-     * @return -1 if disabled, but never null.
+     * @return Actual port number or -1 if disabled.
      */
     public synchronized int getActualPort() {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/main/modules/sshd/SSHD.java]: Nice 2 have: `@CheckForSigned` to enforce checks

File: src/main/java/org/jenkinsci/main/modules/sshd/SSHD.java [hunk 4]
```diff
@@ -64,7 +65,7 @@ public SSHD() {
      * Returns the configured port to run SSHD.
      *
      * @return
-     *      -1 to disable this, 0 to run with a random port, otherwise the port number.
+     *      -1 if disabled, 0 if random port is selected, otherwise the port number configured.
      */
     public int getPort() {
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/main/modules/sshd/SSHD.java]: Nice 2 have: `@CheckForSigned` to enforce checks

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers some reviews of edge cases would be appreciated.

The only risk I see is that SSH-dependent `JenkinsRule`-based tests and Configuratiion Management logic will need to wait and check the SSH status before they will be able to really connect to the instance. Looks fine for SSH clients, which should poll the status anyway

**Comment by olivergondza** [comment]
Given my main motivation here are the tests and if this turns out to cause some problems for the tests, we may want to rethink the startup logic for `jenkins-test-harness` so it will not start the server at all unless someone call `SSHD.get()` as a signal that the test do care. I presume 99+% of the tests does not...

**Comment by oleg-nenashev** [comment]
Regarding JTH, I agree.
The most of the tests should not wait for SSHD startup, and actually SSHD should be disabled in them at all if not required. Maybe JTH could just disable SSHD when running Jenkins and then tests can enable it by setting the SSHD port when it's required

**Comment by olivergondza** [comment]
I guess we will have to try running the core tests against this to see if it would cause problems before merging this.

**Comment by olivergondza** [comment]
@oleg-nenashev, no further reviews, mind if I release this and start using it as a test dependency so we start receiving feedback?

**Comment by oleg-nenashev** [comment]
@olivergondza fine for me

**Comment by jglick** [comment]
Sorry, just saw this and it looks potentially wrong.

The `port` was already defaulted to -1 (as of #12 in `sshd-1.11` ~ https://github.com/jenkinsci/jenkins/pull/2795 in 2.46.2 / 2.54), so `start()` should have returned immediately. Only tests which actually need SSHD should be doing

```java
SSHD.get().setPort(0);
```

So what tests were you seeing stuck @olivergondza? Perhaps you were using an older Jenkins baseline and simply needed to update?

**Comment by jglick** [comment]
IOW this change was not necessary for test performance; and may be introducing a race condition in production environments: you would expect that after Jenkins claims to be “up and running” that you could connect to a configured SSH service, for example via CLI, but now there is a window where you may not be able to.

**Comment by olivergondza** [comment]
> So what tests were you seeing stuck @olivergondza? Perhaps you were using an older Jenkins baseline and simply needed to update?

You are right this was resolved by updating sshd to deploy your fix. I experienced all the tests to be stuck for tens of seconds waiting for Jenkins/SSHD to come up.

**Comment by daniel-beck** [comment]
PR build failure is the flaky protocols warning.
